janusgraph-core/src/main/java/org/janusgraph/graphdb/database/cache/CacheInvalidationService.java

```diff
@@ -0,0 +1,192 @@
+     * the `key` before passing the key to this method.
+     * <p>
+     * In case vertices invalidation is needed by processing transaction logs via {@link org.janusgraph.core.log.ChangeState}
+     * then the method {@link org.janusgraph.core.log.ChangeState#getVertices(Change)} can be used to retrieve all
+     * changed vertices and passing their ids to {@link #markVertexAsExpiredInEdgeStore(Long)}.
+     *
+     * @param key key to expire in `edgeStore` cache
+     */
+    void markKeyAsExpiredInEdgeStore(StaticBuffer key);
+
+    /**
+     * Marks specific key as expired in `indexStore` cache.
+     * It will make sure that any retrieved data associated with this key will be invalidated in index cache.
+     * <p>
+     * Warning! This doesn't invalidate `edgeStore` cache. Thus, trying to return properties or edges for the vertex
+     * may still return stale data. See {@link #markKeyAsExpiredInEdgeStore(StaticBuffer)} JavaDoc to learn how to invalidate
+     * data for `edgeStore`.
+     * <p>
+     * To form a `key` for invalidation it is needed to know vertex id, updated property name, previous value (if there was any),
+     * new value (if there is any) from which IndexUpdate key can be formed.
+     * <p>
+     * Usually this information can be found in a retrieved mutation logs which are passed via {@link org.janusgraph.core.log.ChangeState}
+     * (described in `Transaction Log` documentation of JanusGraph). In case `indexStore` invalidation should triggered
+     * from processing transaction logs via `ChangeState` then it can be done using an example like below:
+     *
+     * <pre>
+     * IndexSerializer indexSerializer = (StandardJanusGraph) graph.getIndexSerializer();
+     * CacheInvalidationService invalidationService = graph.getDBCacheInvalidationService();
+     * changeState.getRelations(Change.ANY).forEach(janusGraphRelation -&gt; {
+     *     if(janusGraphRelation.isProperty() &amp;&amp; janusGraphRelation instanceof InternalRelation){
+     *         JanusGraphVertexProperty property = (JanusGraphVertexProperty) janusGraphRelation;
+     *         if(property.element() instanceof InternalVertex){
+     *             Collection&lt;IndexUpdate&gt; indexUpdates = indexSerializer.getIndexUpdates((InternalVertex) property.element(),
+     *                 Collections.singleton((InternalRelation) property));
+     *
+     *             for(IndexUpdate indexUpdate : indexUpdates){
+     *                 StaticBuffer keyToInvalidate = (StaticBuffer) indexUpdate.getKey();
+     *                 invalidationService.markKeyAsExpiredInIndexStore(keyToInvalidate);
+     *             }
+     *
+     *             invalidationService.forceClearExpiredKeysInIndexStoreCache();
+     *             invalidationService.forceInvalidateVertexInEdgeStoreCache(property.element().longId());
+     *         }
+     *     }
+     * });
```

Comment:
This example only partially correct because we also need to:
- Check that key is actually a `StaticBuffer` (for composite index) and not a `String` (for mixed index).
- Check for edge index updates
- Check added / removed vertices

If I add all these checks to this example than it becomes too complex to read in JavaDoc. Instead of this I propose removing this example and when #3155 is finished to simply add references to the appropriate methods for the developers who are interested to read it.

Reply:
Updated JavaDoc